Runtime: Don't append local discriminators to locally defined type names.

5at5ish opened the pull request:
If a type was defined locally, a printed representation of an instance of that
type would contain a local discriminator. Introducing an additional demangling
option allows to opt out of this behaviour.

Resolves [SR-6787](https://bugs.swift.org/browse/SR-6787).

jrose-apple commented on include/swift/Demangling/Demangle.h:
```diff
@@ -67,6 +68,7 @@ struct DemangleOptions {
     Opt.ShortenThunk = true;
     Opt.ShortenValueWitness = true;
     Opt.ShortenArchetype = true;
+    Opt.ShowLocalDiscriminators = true;
```
Shouldn't this one be `false`, since ShowPrivateDiscriminators and DisplayModuleNames are also `false`?

5at5ish replied:
[This](https://github.com/apple/swift/blob/master/test/SILGen/generic_local_property.swift#L24) check fails if it's `false`. We could customize options [here](https://github.com/apple/swift/blob/master/lib/SIL/SILPrinter.cpp#L74), to address that case.

jrose-apple replied:
I think I would just pass `-Xllvm -sil-full-demangle` for that one test.

jrose-apple commented on stdlib/public/runtime/Casting.cpp:
```diff
@@ -100,6 +100,7 @@ static void _buildNameForMetadata(const Metadata *type,
 
   Demangle::DemangleOptions options;
   options.QualifyEntities = qualified;
+  options.ShowLocalDiscriminators = false;
```
I think this one probably goes with `qualified`, like `ShowPrivateDiscriminators`. Because…

jrose-apple commented on test/stdlib/StringDescribing.swift:
```diff
@@ -22,4 +22,19 @@ StringDescribingTestSuite.test("String(reflecting:) should include extra stuff i
+  expectEqual(String(describing: type(of: t)), "LocalType")
+}
+
+StringDescribingTestSuite.test("String(reflecting:) shouldn't include local discriminator") {
```
…`String(reflecting:)` should give a unique answer for every unique type.

5at5ish replied:
Should assertion message be reworded?

jrose-apple replied:
Yes, like the one for private discriminators.

jrose-apple reviewed (changes requested):
Thanks for picking this up, Timur!

5at5ish committed:
Runtime: Don't append local discriminators to locally defined type names.

If a type was defined locally, a printed representation of an instance of that
type would contain a local discriminator. Introducing an additional demangling
option allows to opt out of this behaviour.

Resolves: SR-6787.

5at5ish commented:
@jrose-apple Addressed your comments.

jrose-apple commented on test/stdlib/StringDescribing.swift:
```diff
@@ -22,4 +22,19 @@ StringDescribingTestSuite.test("String(reflecting:) should include extra stuff i
+  let representation = String(reflecting: type(of: t))
+
+  expectEqual(representation.prefix(6), "main.(")
+  expectEqual(representation.suffix(11), ").LocalType")
```
Ah, now we should be checking that the "#1" *is* there. Doesn't this fail?

5at5ish replied:
It doesn't fail. Isn't it expected behaviour for locally defined types?

jrose-apple replied:
Hm, no. If `String(reflecting:)` uniquely identifies a type, that means it has to include the "`#1`". Maybe look for where else ShowPrivateDiscriminators is set to true? Or just combine the two settings, since I can't think why we'd ever want them to be different.

5at5ish replied:
It looks like after merging [this](https://github.com/apple/swift/pull/13468/files#diff-360e359aecc8c097fac060c301c6d378R1005) change, the logic should be reconsidered.

jrose-apple replied:
I don't think that affects this.

5at5ish replied:
Added a comment to the ticket.

CodaFi commented:
There's been a change in the descriptor format and part of this was addressed by #12298.  I'm going to close this out.  Thank you for taking a look at this issue!
